### PR TITLE
Update GQL schema

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR updates the GQL schema for Recurrence and also hooks into core to update GQL definitions.
See [#3021](https://github.com/eventespresso/event-espresso-core/pull/3021)

### CODE REVIEW

File: src/domain/services/graphql/ExtendCore.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * @param GraphQLFieldInterface[] $fields
+     * @return GraphQLFieldInterface[]
+     * @since $VID:$
+     */
+    public function registerCoreDatetimeFields($fields)
+    {
+        $newFields = [
+            // add recurrence field to datetime mutation input
```
Comment (nicolad) [inline src/domain/services/graphql/ExtendCore.php:57]: cool :+1:

File: src/domain/services/graphql/ExtendCore.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+     * @return array
+     * @since $VID:$
+     */
+    public function addDatetimeConnectionWhereParams($where_params, $source, $args)
+    {
+        $arg_mapping = [
+            'recurrence'     => 'RCR_ID',
+            'recurrenceIn'   => 'RCR_ID',
+            'recurrenceId'   => 'RCR_ID',
+            'recurrenceIdIn' => 'RCR_ID',
```
Comment (nicolad) [inline src/domain/services/graphql/ExtendCore.php:134]: dumb question: should each of the items from the mapping point to the same string?

Comment (tn3rb) [reply src/domain/services/graphql/ExtendCore.php:134]: yes. they are used in different kinds of queries but ultimately all point to the same field.

for example, some queries might be something like: 

- select the recurrence with the `RCR_ID` ###
- select the recurrences with an `RCR_ID` in [###, ###, ###, ###]
- select the datetime with a relation to the recurrence with the `RCR_ID` ###
- select the datetimes with relations to the recurrences with an `RCR_ID` in [###, ###, ###, ###]

File: src/domain/services/graphql/ExtendCore.php [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                        if (is_string($value)) {
+                            $value = sanitize_text_field($value);
+                        }
+                        return $value;
+                    },
+                    $value
+                );
+            } elseif (is_string($value)) {
+                $value = sanitize_text_field($value);
+            }
```
Comment (tn3rb) [inline src/domain/services/graphql/ExtendCore.php:154]: we should maybe think about extracting this into some kind of GraphQlFieldValidation class that can be shared between core and add-ons

File: src/domain/services/graphql/ExtendCore.php [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+     * @param string|string[] $ID
+     * @return mixed
+     */
+    protected function convertGlobalId($ID)
+    {
+        if (is_array($ID)) {
+            return array_map([ $this, 'convertGlobalId' ], $ID);
+        }
+        $parts = Relay::fromGlobalId($ID);
+        return ! empty($parts['id']) ? $parts['id'] : null;
```
Comment (tn3rb) [inline src/domain/services/graphql/ExtendCore.php:181]: have seen this logic scattered all over the place which would make another good candidate for extraction.

beyond the scope of this PR though

File: src/domain/services/graphql/ExtendCore.php (hunk 2 above)
Comment (manzoorwanijk) [reply src/domain/services/graphql/ExtendCore.php:134]: A minor correction to @tn3rb's explanation:
- `recurrence` refers to GUID
- `recurrenceId` refers to DBID

It supports both the IDs.

File: src/domain/services/graphql/ExtendCore.php (hunk 4 above)
Comment (manzoorwanijk) [reply src/domain/services/graphql/ExtendCore.php:181]: 💯

File: src/domain/services/graphql/RegisterSchema.php [hunk 5]
```diff
@@ -134,27 +134,29 @@ public function addDatetimeConnectionWhereParams($where_params, $source, $args)
             'recurrenceIdIn' => 'RCR_ID',
         ];
         $id_fields = ['recurrence', 'recurrenceIn'];
-        foreach ($args['where'] as $arg => $value) {
-            if (! array_key_exists($arg, $arg_mapping)) {
-                continue;
+        if (! empty($args['where'])) {
```
Comment (tn3rb) [inline src/domain/services/graphql/RegisterSchema.php:137]: 👍

**Pull request merged**